# Data Profiling

## Project Overview

### Business Scenario
A retail analytics team has received web-scraped pricing data from five major UK supermarkets. Before the data can be used for reporting, price comparisons, or business analysis, its quality must be assessed to identify any issues that could affect the reliability of the analysis.

### Objective
The objective of this phase is to understand the structure and quality of the datasets before performing any cleaning. This ensures that any changes made during the cleaning process are based on evidence rather than assumptions.

---

## What We Know

Based on the available information, we know that:

- The data has been collected through web scraping.
- There are five separate datasets, one for each supermarket.
- The data contains daily supermarket prices.
- Each dataset includes product names, prices, units, categories, dates, and an own-brand indicator.
- Web-scraped data may contain inconsistencies or quality issues that need to be investigated.

---

## What We Need to Investigate

Before cleaning the data, the following questions need to be answered.

### Completeness
- Are there any missing values?
- Are there empty strings?
- Are important fields incomplete?

### Validity
- Are prices stored as numeric values?
- Are dates in a valid format?
- Are unit values valid?
- Are there any impossible prices?

### Consistency
- Are category names consistent?
- Are units written consistently?
- Are boolean values stored consistently?
- Are there differences in spelling or capitalisation?

### Uniqueness
- Are there duplicate rows?
- Are duplicate products expected?

### Timeliness
- Do all records contain scrape dates?
- Are there unexpected gaps in the dates?

---

## Investigation Plan

The data profiling stage will answer each of these questions before any cleaning is performed. The findings will guide the data cleaning and validation stages of the project.

# UK Supermarket Data Quality Assessment

## Project Overview

This project analyses pricing data from five UK supermarkets.

The purpose of this notebook is to profile the raw datasets before cleaning them.

In [4]:
import pandas as pd 


In [5]:
aldi = pd.read_csv("../data/Raw/All_Data_Aldi.csv")

In [3]:
aldi.head() 

,supermarket,prices_(£),prices_unit_(£),unit,names,date,category,own_brand
0,Aldi,3.09,0.14,unit,Mamia Ultra-fit Peppa Pig Nappy Pants 22 Pack/...,20240413,baby_products,False
1,Aldi,3.09,0.17,unit,Mamia Ultra-fit Peppa Pig Nappy Pants 18 Pack/...,20240413,baby_products,False
2,Aldi,3.59,0.09,unit,Mamia Ultra-fit Nappy Pants 40 Pack/Size 4,20240413,baby_products,False
3,Aldi,4.79,0.32,unit,Mamia Boy's Night Pants 15 Pack,20240413,baby_products,False
4,Aldi,4.79,0.32,unit,Mamia Girl's Night Pants 15 Pack,20240413,baby_products,False


In [6]:
aldi.info()

<class 'pandas.DataFrame'>
RangeIndex: 464863 entries, 0 to 464862
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   supermarket      464863 non-null  str    
 1   prices_(£)       464863 non-null  float64
 2   prices_unit_(£)  464859 non-null  float64
 3   unit             464859 non-null  str    
 4   names            464863 non-null  str    
 5   date             464863 non-null  int64  
 6   category         464863 non-null  str    
 7   own_brand        464863 non-null  bool   
dtypes: bool(1), float64(2), int64(1), str(4)
memory usage: 25.3 MB


In [7]:
aldi.shape

(464863, 8)

In [8]:
aldi.columns

Index(['supermarket', 'prices_(£)', 'prices_unit_(£)', 'unit', 'names', 'date',
       'category', 'own_brand'],
      dtype='str')

In [10]:
aldi.isnull().sum()

supermarket        0
prices_(£)         0
prices_unit_(£)    4
unit               4
names              0
date               0
category           0
own_brand          0
dtype: int64

## Missing Values Summary

The missing value assessment identified only two columns containing null values:

- `prices_unit_(£)` contains 4 missing values.
- `unit` contains 4 missing values.

All other columns are complete.

At this stage, no cleaning action has been taken. The missing records will be investigated later to determine whether they represent genuine data issues or expected missing information.

In [11]:
aldi.duplicated().sum()


np.int64(0)

## Duplicate Rows Summary

A check for duplicate rows found **0 duplicate records**.

This indicates that there are no completely identical rows in the Aldi dataset. At this stage, no duplicate records need to be removed.

It is important to note that this check only identifies identical rows. Duplicate product names may still exist and will be investigated separately, as products can legitimately appear multiple times across different dates.

In [12]:
aldi["date"].head()

0    20240413
1    20240413
2    20240413
3    20240413
4    20240413
Name: date, dtype: int64

## Date Column Assessment

The `date` column is currently stored as an integer (`int64`).

A review of the values shows that the dates follow the `YYYYMMDD` format (for example, `20240413` represents **13 April 2024**).

Although the values are valid, the column is not stored as a datetime data type. During the data cleaning phase, it will be converted to a datetime format to support filtering, grouping, and time-based analysis.

In [13]:
aldi["category"].value_counts()

category
fresh_food         154187
food_cupboard      118609
drinks              57463
frozen              44264
household           25549
health_products     21616
bakery              20540
pets                 9559
baby_products        7914
free-from            5162
Name: count, dtype: int64

## Category Assessment

The dataset contains **10 unique product categories**.

The category names appear to follow a consistent naming convention with no obvious differences in spelling or capitalisation.

One observation is that most category names use underscores (for example, `fresh_food` and `food_cupboard`), while `free-from` uses a hyphen. This will be reviewed during the cleaning phase to determine whether standardisation is necessary.

At this stage, no category inconsistencies requiring correction have been identified.

# Data Profiling Summary

## Findings

- The dataset contains **464,863 rows** and **8 columns**.
- All expected columns are present.
- Most data types are appropriate for analysis.
- The `date` column is stored as an integer and should be converted to a datetime format during cleaning.
- Only **4 missing values** were identified in the `prices_unit_(£)` and `unit` columns.
- No duplicate rows were found.
- The dataset contains **10 product categories** with no obvious naming inconsistencies.

## Next Steps

The next phase of the project will investigate:

- Missing records
- Duplicate product names
- Date conversion
- Data validation
- Data cleaning decisions

In [15]:
aldi["names"].duplicated().sum()


np.int64(457601)

In [16]:
aldi["names"].nunique()

7262

In [17]:
aldi["names"].value_counts().head(10)

names
Mamia Newborn Nappies 24 Pack/Size 1 2-5kg/4-11lbs         92
Mamia Newborn Nappies 60 Pack/Size 2 Mini 3-6kg/6-13lbs    92
The Foodie Market Exotic Coconut Mix 25g                   92
Dominion Leo Lion Fruit Jellies 170g                       92
Dominion Walter Worm Chewy Fruit Gums 170g                 92
Dominion Wendy Worm Chewy Fruit Gums 170g                  92
The Foodie Market Whole Chia Seeds 200g                    92
The Foodie Market Dried Prunes 500g                        92
Foodie Market Mixed Nuts 200g                              92
Foodie Market Brazil Nuts 200g                             92
Name: count, dtype: int64

## Product Name Assessment

The dataset contains **7,262 unique product names**.

Several products appear repeatedly throughout the dataset. For example, the most frequently occurring products each appear **92 times**.

This behaviour is expected because the dataset contains daily web-scraped observations. The repeated product names most likely represent the same products being collected on different dates rather than duplicate records.

Therefore, repeated product names are considered valid and do not represent a data quality issue.


In [18]:
(aldi["prices_unit_(£)"] <=0).sum()

np.int64(2093)

In [19]:
(aldi["prices_unit_(£)"] <0).sum()

np.int64(0)

In [20]:
(aldi["prices_unit_(£)"] ==0).sum()

np.int64(2093)

In [21]:
aldi["prices_unit_(£)"].describe()

count    464859.000000
mean          7.993101
std          41.288759
min           0.000000
25%           2.230000
50%           5.300000
75%           9.000000
max        2343.800000
Name: prices_unit_(£), dtype: float64

## Price Assessment

The dataset contains **no negative prices**, which indicates that invalid negative values are not present.

However, **2,093 records contain a price of £0.00**. At this stage, these records have not been removed because they may represent valid business scenarios such as unavailable products or promotional items.

The descriptive statistics also indicate that the maximum price per unit is relatively high. This value will be investigated further during the validation stage to determine whether it represents a genuine product or a potential outlier.

In [22]:
aldi["prices_(£)"].describe()

count    464863.000000
mean          2.216073
std           2.323695
min           0.010000
25%           0.990000
50%           1.590000
75%           2.490000
max         383.000000
Name: prices_(£), dtype: float64

## Selling Price Assessment

The `prices_(£)` column contains no negative or zero values. The minimum recorded selling price is **£0.01**, indicating that all products have a positive selling price.

The median selling price is **£1.59**, suggesting that most products are relatively low-cost, which is consistent with supermarket retail data.

The highest recorded selling price is **£383.00**. This value will be investigated further during the validation stage to determine whether it represents a legitimate high-value product or a potential outlier.

In [23]:
aldi[aldi["prices_unit_(£)"].isnull()]

,supermarket,prices_(£),prices_unit_(£),unit,names,date,category,own_brand
337885,Aldi,0.65,NaN,NaN,Saxon Biodegradable Toilet Tissue Wipes - Fine...,20240205,household,False
342073,Aldi,0.65,NaN,NaN,Saxon Biodegradable Toilet Tissue Wipes - Fine...,20240204,household,False
455796,Aldi,0.65,NaN,NaN,Saxon Biodegradable Toilet Tissue Wipes - Fine...,20240110,household,False
461162,Aldi,0.65,NaN,NaN,Saxon Biodegradable Toilet Tissue Wipes - Fine...,20240109,household,False


In [24]:
aldi[aldi["unit"].isnull()]

,supermarket,prices_(£),prices_unit_(£),unit,names,date,category,own_brand
337885,Aldi,0.65,NaN,NaN,Saxon Biodegradable Toilet Tissue Wipes - Fine...,20240205,household,False
342073,Aldi,0.65,NaN,NaN,Saxon Biodegradable Toilet Tissue Wipes - Fine...,20240204,household,False
455796,Aldi,0.65,NaN,NaN,Saxon Biodegradable Toilet Tissue Wipes - Fine...,20240110,household,False
461162,Aldi,0.65,NaN,NaN,Saxon Biodegradable Toilet Tissue Wipes - Fine...,20240109,household,False


## Investigation of Missing Values

The four missing values identified in the `prices_unit_(£)` and `unit` columns were investigated further.

All missing values belong to the same product:

- **Saxon Biodegradable Toilet Tissue Wipes**

The product appears on four different scrape dates, and each occurrence is missing the same two fields.

This suggests that the missing values are product-specific rather than representing a broader issue with the dataset.

No rows will be removed during the profiling stage. The handling of these missing values will be considered during the data cleaning phase.

In [4]:
import pandas as pd

aldi = pd.read_csv("../data/raw/All_Data_Aldi.csv")

asda = pd.read_csv("../data/raw/All_Data_ASDA.csv")

morrisons = pd.read_csv("../data/raw/All_Data_Morrisons.csv")

sains = pd.read_csv("../data/raw/All_Data_Sains.csv")

tesco = pd.read_csv("../data/raw/All_Data_Tesco.csv")

C:\Users\Atq12\AppData\Local\Temp\ipykernel_22748\3819125500.py:5: DtypeWarning: Columns (0: own_brand) have mixed types. Specify dtype option on import or set low_memory=False.
  asda = pd.read_csv("../data/raw/All_Data_ASDA.csv")


In [30]:
def profile_dataset(df, name):
    ...

In [33]:
profile_dataset(aldi, "Aldi")

Dataset: Aldi
Rows: 464,863
Columns: 8

Missing Values
supermarket        0
prices_(£)         0
prices_unit_(£)    4
unit               4
names              0
date               0
category           0
own_brand          0
dtype: int64

Duplicate Rows
0

Unique Categories
10

Date Range
20240109 to 20240413


In [34]:
profile_dataset(asda, "ASDA")

Dataset: ASDA
Rows: 2,456,414
Columns: 8

Missing Values
supermarket          0
prices_(£)           7
prices_unit_(£)    199
unit               199
names               26
date                 0
category             0
own_brand           26
dtype: int64

Duplicate Rows
0

Unique Categories
11

Date Range
20240109 to 20240413


In [35]:
profile_dataset(morrisons, "Morrisons")

Dataset: Morrisons
Rows: 1,794,065
Columns: 8

Missing Values
supermarket          0
prices_(£)           0
prices_unit_(£)    110
unit               110
names                0
date                 0
category             0
own_brand            0
dtype: int64

Duplicate Rows
0

Unique Categories
11

Date Range
20240109 to 20240413


In [36]:
profile_dataset(sains, "Sainsbury's")

Dataset: Sainsbury's
Rows: 2,600,289
Columns: 8

Missing Values
supermarket          0
prices_(£)           0
prices_unit_(£)    209
unit               209
names                0
date                 0
category             0
own_brand            0
dtype: int64

Duplicate Rows
0

Unique Categories
11

Date Range
20240109 to 20240413


In [37]:
profile_dataset(tesco, "Tesco")

Dataset: Tesco
Rows: 2,213,611
Columns: 8

Missing Values
supermarket        0
prices_(£)         0
prices_unit_(£)    2
unit               2
names              0
date               0
category           0
own_brand          0
dtype: int64

Duplicate Rows
23828

Unique Categories
10

Date Range
20240109 to 20240413


# Investigation 1: Tesco Duplicate Rows

The initial profiling identified **23,828 duplicate rows** in the Tesco dataset.

Before removing any records, this investigation aims to determine:

- Whether the duplicate rows are exact duplicates.
- How many duplicate records exist.
- Whether removing them is justified.

In [5]:
tesco.duplicated().sum()    

np.int64(23828)

In [6]:
tesco[tesco.duplicated()].head(10)

,supermarket,prices_(£),prices_unit_(£),unit,names,date,category,own_brand
2070636,Tesco,2.45,14.60,kg,Soreen Lift Bars Raspberry & Vanilla 4 Pack 168G,20240115,bakery,False
2070637,Tesco,1.45,1.80,kg,Tesco Crusty White Sandwich Loaf Sliced 800G,20240115,bakery,True
2070638,Tesco,1.60,0.40,unit,Baker Street 4 Hot Dog Rolls,20240115,bakery,False
2070639,Tesco,2.00,4.40,kg,Jason's Sourdough Loaf Majestic Malted 450G,20240115,bakery,False
2070640,Tesco,1.35,5.60,kg,Crosta & Mollica 6 Durum Wheat Piadina Midi Fl...,20240115,bakery,False
2070641,Tesco,12.50,3.79,l,Old Mout Cider Strawberry & Apple Can 10x330ml,20240115,drinks,False
2070642,Tesco,2.00,0.88,l,Tesco Organic British Semi Skimmed Milk 2.272L...,20240115,drinks,True
2070643,Tesco,0.40,2.00,l,Tesco Apple Juice From Concentrate 200Ml,20240115,drinks,True
2070644,Tesco,4.10,3.00,l,Innocent Orange Juice Smooth 1.35 Litre,20240115,drinks,False
2070645,Tesco,4.00,2.00,l,Lucozade Sport Orange 4X500ml,20240115,drinks,False


In [7]:
tesco[tesco.duplicated(keep=False)].sort_values(by=["names","date"]).head(20)

,supermarket,prices_(£),prices_unit_(£),unit,names,date,category,own_brand
2065914,Tesco,2.00,2.00,unit,1 Marigold Extra Life Gloves Kitchen Large,20240115,household,False
2089742,Tesco,2.00,2.00,unit,1 Marigold Extra Life Gloves Kitchen Large,20240115,household,False
2047686,Tesco,2.00,2.00,unit,1 Marigold Extra Life Gloves Kitchen Medium,20240115,household,False
2071514,Tesco,2.00,2.00,unit,1 Marigold Extra Life Gloves Kitchen Medium,20240115,household,False
2051636,Tesco,8.99,2.25,unit,10 Motives Refills Menthol 16Mg,20240115,home,False
2075464,Tesco,8.99,2.25,unit,10 Motives Refills Menthol 16Mg,20240115,home,False
2058519,Tesco,8.99,2.25,unit,10 Motives Refills Regular 16Mg,20240115,home,False
2082347,Tesco,8.99,2.25,unit,10 Motives Refills Regular 16Mg,20240115,home,False
2048458,Tesco,9.99,9.99,unit,10 Motives V2 Chargeable Regular,20240115,home,False
2072286,Tesco,9.99,9.99,unit,10 Motives V2 Chargeable Regular,20240115,home,False


## Findings

The duplicate rows in the Tesco dataset were investigated by comparing the original records with their corresponding duplicates.

The sampled records were identical across all columns, including:

- Supermarket
- Selling price
- Price per unit
- Unit
- Product name
- Date
- Category
- Own-brand indicator

Based on this investigation, the duplicate rows appear to be exact duplicate records rather than separate observations.

No records have been removed during the profiling stage. Duplicate removal will be performed and documented in the data cleaning notebook.

In [8]:
tesco[tesco.duplicated(keep=False)].groupby(list(tesco.columns)).size().value_counts()

2    23828
Name: count, dtype: int64

## Conclusion

The Tesco dataset contains **23,828 duplicate rows**.

Further investigation showed that:

- The duplicate records are identical across all columns.
- Each duplicated record appears exactly **twice**.
- No evidence was found of records appearing more than twice.

Based on these findings, the duplicate rows are considered true duplicate records rather than valid repeated observations.

The duplicate rows will be removed during the data cleaning stage. No records are removed during the profiling notebook.

# Investigation 2: ASDA Missing Product Names

In [9]:
asda[asda["names"].isnull()]

,supermarket,prices_(£),prices_unit_(£),unit,names,date,category,own_brand
212170,ASDA,NaN,NaN,NaN,NaN,20240406,fresh_food,NaN
339302,ASDA,NaN,NaN,NaN,NaN,20240401,household,NaN
477644,ASDA,NaN,NaN,NaN,NaN,20240327,free-from,NaN
555891,ASDA,1.40,6.36,kg,NaN,20240324,food_cupboard,NaN
592865,ASDA,1.40,6.36,kg,NaN,20240322,food_cupboard,NaN
1338635,ASDA,NaN,NaN,NaN,NaN,20240222,free-from,NaN
1783046,ASDA,NaN,NaN,NaN,NaN,20240205,fresh_food,NaN
1863923,ASDA,1.29,3.90,l,NaN,20240201,food_cupboard,NaN
1870373,ASDA,1.29,3.90,l,NaN,20240131,food_cupboard,NaN
1896608,ASDA,1.29,3.90,l,NaN,20240130,food_cupboard,NaN


## Findings

The ASDA dataset contains **26 records with missing product names**.

Investigation shows that the missing records fall into two groups:

1. Records where almost all product information is missing, suggesting incomplete web-scraped observations.
2. Records where pricing information is present but the product name is missing.

These records will be investigated further during the cleaning stage before any removal or imputation decisions are made.

In [10]:
asda[asda["names"].isnull()].isnull().sum()

supermarket         0
prices_(£)          7
prices_unit_(£)     7
unit                7
names              26
date                0
category            0
own_brand          26
dtype: int64

## Conclusion

The ASDA dataset contains **26 records with missing product names**.

The investigation identified two distinct groups:

- **7 records** are largely incomplete, with missing product name, selling price, unit price, unit, and own-brand information. These appear to be incomplete web-scraped records.

- **19 records** contain valid pricing and unit information but are missing only the product name and own-brand fields. These records retain useful information and require further investigation before any cleaning decision is made.

No records have been modified or removed during the profiling stage. Appropriate handling of these records will be documented in the data cleaning notebook.

# Investigation 3: Category Consistency

The initial profiling showed that Aldi and Tesco contain 10 product categories, while ASDA, Morrisons and Sainsbury's contain 11.

This investigation compares the category names across supermarkets to determine whether the difference represents a genuine business difference or a data quality issue.

In [11]:
sorted(aldi["category"].unique())

['baby_products',
 'bakery',
 'drinks',
 'food_cupboard',
 'free-from',
 'fresh_food',
 'frozen',
 'health_products',
 'household',
 'pets']

In [12]:
sorted(asda["category"].unique())

['baby_products',
 'bakery',
 'drinks',
 'food_cupboard',
 'free-from',
 'fresh_food',
 'frozen',
 'health_products',
 'home',
 'household',
 'pets']

## Initial Findings

Comparison of the category values shows that Aldi contains **10 categories**, while ASDA contains **11 categories**.

The additional category identified in the ASDA dataset is:

- `home`

At this stage, there is no evidence that this represents a data quality issue. It may reflect differences in product categorisation between supermarkets rather than an error in the dataset.

Further comparison with the remaining supermarkets will determine whether the `home` category is consistently used.

In [13]:
sorted(morrisons["category"].unique())

['baby_products',
 'bakery',
 'drinks',
 'food_cupboard',
 'free-from',
 'fresh_food',
 'frozen',
 'health_products',
 'home',
 'household',
 'pets']

In [14]:
sorted(sains["category"].unique())

['baby_products',
 'bakery',
 'drinks',
 'food_cupboard',
 'free-from',
 'fresh_food',
 'frozen',
 'health_products',
 'home',
 'household',
 'pets']

## Conclusion

Category values were compared across all five supermarkets.

Results showed:

- Aldi and Tesco each contain **10 categories**.
- ASDA, Morrisons and Sainsbury's each contain **11 categories**, including an additional category named `home`.

The `home` category appears consistently across three supermarkets, suggesting that it represents a legitimate product classification rather than a data quality issue.

No category names will be modified during the data cleaning stage. The original categorisation used by each supermarket will be preserved.